app/models/job_step.rb

```diff
@@ -66,18 +86,58 @@ def add_execution(node)
     Execution.new(
       id: "#{self.job.id}.#{id}.#{node.name}",
       job_step: self,
-      node: node,
+      node_name: node.name,
     ).tap do |execution|
       self.executions << execution
     end
   end
 
+  def cleanup
+    if Async::Task.current?
+      Async::IO::Threads.new.async { FileUtils.rm_rf(env_path) }.wait
+    else
+      FileUtils.rm_rf(env_path)
+    end
+  end
+
   def display_id
     "#{job.display_id}.#{id}"
   end
 
+  def env
+    # We deliberately don't cache the value here.
+    return @env if @env
+    serialized_env = 
+      if Async::Task.current?
+        Async::IO::Threads.new.async { File.read(env_path) }.wait
+      else
+        File.read(env_path)
+      end
+    Hash[serialized_env.split("\0").map { |pairs| pairs.split('=', 2) }]
+  end
+
   def execution_for(node_name)
-    executions.detect { |exe| exe.node.name == node_name }
+    executions.detect { |exe| exe.node_name == node_name }
+  end
+
+  def write
+    Async::IO::Threads.new.async do
+      FileUtils.mkdir_p(dirname)
+      serialized_env = env.map { |k, v| "#{k}=#{v}" }.join("\0")
+      File.write(env_path, serialized_env)
```

Comment:
My intention with this piece of code was to make the write atomic.  So that we would never get half of the registry written to the file.  I hadn't considered making the writes ordered.  You're right, we ought to do that too.

I can't see how interleaved reads and writes could cause any issues.  The backup is created with a hardlink.  This is an atomic operation.  The new version of the file is created with a rename.  This is also an atomic operation.  In both cases they cannot interleave with a read in any way that could cause partial or inconsistent data to be read.

Old data could be read.  That is worked around by only loading the registries when the application loads, prior to it accepting any connections or its periodic timer being started.  If this changes, we may need to prevent old data from being read.

Reply:
The above comment was actually talking about `Persistence` not `JobStep`.

Regarding `JobStep`, the code is safe because it is called only once.  That can be made clearer.